Refactor ResourceLayer and implement proper TS Tiberium and Veins logic

pchote opened the pull request:
![](https://user-images.githubusercontent.com/167819/105781555-b1daa300-5f6a-11eb-9459-2f071ccf97c2.png)

This PR tackles the next ugly piece of legacy OpenRA code blocking TS progress. The first commits work through cleaning up the existing logic, and should be relatively straightforward to follow when reading each commit in order. The final commit uses the improved logic to implement the proper vein and tiberium logic in TS. I can split these commits over a series of smaller PRs if anyone feels this is too much to review in one go.

Some points of note:
* The hardcoded `ResourceType` trait has been removed, and resources definitions now work like most other things in the engine: string labels that reference data the traits that use them. This fixes most of the modding limitations by allowing mods to subclass or replace traits instead of having to ask for changes to `ResourceType` and the map editor.
* The hardcoded assumption that resources always and only provide money has also been removed: it is now the responsibility of `IAcceptResources` to decide what happens, and the resource values are now defined on `Refinery`.
* TS has some horribly specific rules about where resources are allowed<sup>1</sup>, which is impossible to generically expose via yaml. This has driven a change in design where the old `Allow*` options have been removed, and mods instead subclass the `(Editor)ResourceLayer` and override methods.
* Mods can now disable the automatic neighbour recalculation and use the density saved in the map data. This is done in TS, but requires an improved resource placement tool in the map editor before we can think about changing the others.

<sup>1</sup>The original TS engine implemented resources as overlays (along with walls, bridges, crates, etc), which meant that harvestable resources were automatically not allowed to overlap with visual only things. Veins were a bit weird in that the LAT border cells were not themselves harvestable (they gave a harvest cursor but the harvester would go to the closest interior cell), which means that our `ResourceLayer` code needs to reserve an empty cell between veins and tiberium so that the `ResourceRenderer` can draw the borders without it overlapping with tiberium. Tiberium and Veins are also allowed on slopes, but only some slopes, and Veins on slopes *don't* have borders. This sounds complicated, but is IMO quite straightforward to manage with the new plumbing.

pchote committed:
Add resource updating methods to IResourceLayer.

pchote committed:
Use IResourceLayer for editor resources.

pchote committed:
Use IResourceLayer in BaseBuilderBotModule.

pchote committed:
Remove custom D2k spice variant logic and improve editor cursor sprite.

pchote committed:
Replace ResourceType with strings in interfaces/public methods.

pchote commented on OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs:
```diff
@@ -89,57 +126,142 @@ public void UpdateCell(CPos cell)
+			var terrainInfo = Map.Rules.TerrainInfo;
+			var terrainType = terrainInfo.TerrainTypes[terrainInfo.GetTerrainInfo(Map.Tiles[cell]).TerrainType].Type;
+
+			// TODO: Check against actors in the EditorActorLayer
```
Note that these TODOs are added as documentation for (non-)existing editor limitations, and are quite out of scope for this PR.

MustaphaTR commented on OpenRA.Mods.Common/UpdateRules/Rules/20201213/ReplaceResourceValueModifiers.cs:
```diff
@@ -0,0 +1,41 @@
+		{
+			get
+			{
+				return "The HarvesterResourceMultiplier trait has been removed, and the RefineryResourceMultiplier trait renamed to ResourceValueMultiplier.";
```
Graion notified me of this change on Discord, while i don't care much about GenAlpha anymore. The HarvesterResourceMultiplier was used for the implementation of Worker Shoes update, which increased the resource dropped by Workers to any Supply Center. Removing the trait would remove this ability, and while i haven't looked in detail, doesn't seem like new trait is able to do that. On the opposite end for RefineryResourceMultiplier was used by Supply Lines upgrade, which increased resource dropped to American Supply Center by any supply unit.

I think rather than keeping 2 traits for this. Adding an actor type check to ResourceValueMultiplier could let us keep this ability. So the trait would sit on the Supply Center on both cases, but you can filter the Worker Shoes case so only modifier only apply to resources dropped by Workers.

pchote replied:
~`INotifyResourceAccepted` should already have access to everything you need to implement that~, otherwise you could subclass Harvester and have that drop a different resource type that the supply depot has a different value for... or just drop more units of resource than they picked up. The driving motivation of this rewrite was to remove all the weird hardcoded behaviours and have mods implement everything based on resource types after all!

pchote replied:
Updated to make `Harvester.TickUnload` virtual, so Generals Alpha can do whatever it needs.

phrohdoh commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -152,7 +152,7 @@ public interface INotifyDelivery { void IncomingDelivery(Actor self); void Deliv
 
 	[RequireExplicitImplementation]
-	public interface INotifyResourceAccepted { void OnResourceAccepted(Actor self, Actor refinery, int amount); }
+	public interface INotifyResourceAccepted { void OnResourceAccepted(Actor self, Actor refinery, string resourceType, int amount, int value); }
```
It would be nice to have documentation explaining the difference between `amount` and `value`.

pchote replied:
Agreed that `amount` is ambiguous, and outside our usual naming scheme. I renamed it to `count` to be clearer.

teinarss commented on OpenRA.Mods.Cnc/Traits/World/TSTiberiumRenderer.cs:
```diff
@@ -0,0 +1,92 @@
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
```
Wrong mod :p

teinarss commented on OpenRA.Mods.Cnc/Traits/World/TSVeinsRenderer.cs:
```diff
@@ -0,0 +1,335 @@
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
```
!

pchote commented:
Fixed.

reaperrr commented on OpenRA.Mods.Cnc/Traits/ResourcePurifier.cs:
```diff
@@ -56,12 +56,12 @@ protected override void Created(Actor self)
 		}
 
-		void INotifyResourceAccepted.OnResourceAccepted(Actor self, Actor refinery, int amount)
+		void INotifyResourceAccepted.OnResourceAccepted(Actor self, Actor refinery, string resourceType, int amount, int value)
```
This should also use `count` instead of `amount` now, right?

reaperrr commented on mods/ts/rules/world.yaml:
```diff
@@ -198,47 +198,39 @@
-	ResourceRenderer:
-		RenderTypes: Tiberium, BlueTiberium, Veins
+	TSTiberiumRenderer:
+		Resources:
```
Nit: Wouldn't it be more consistent to call this `ResourceTypes`, since you use `ResourceType` on the VeinsRenderer (and internally)?

pchote replied:
My logic here was that `ResourceType` refers to the string label, whereas these definitions are listing the resource info for a given type which make them a bit different. I don't mind changing this if you or others still prefer `ResourceTypes`.

reaperrr replied:
OK, makes sense. I don't really have a preference here, so as far as I'm concerned we can leave it like that then.

pchote replied:
After more thought I think I will change this.

reaperrr commented on mods/ts/rules/world.yaml:
```diff
@@ -278,7 +270,24 @@ World:
-	ResourceLayer:
+	TSResourceLayer:
+		VeinholeActors: veinhole
+		Resources:
```
Same here...

reaperrr commented on mods/ts/rules/world.yaml:
```diff
@@ -376,7 +385,27 @@ EditorWorld:
 	EditorCursorLayer:
-	EditorResourceLayer:
+	TSEditorResourceLayer:
+		Resources:
```
...and here?

teinarss commented on OpenRA.Mods.Cnc/Traits/World/TSResourceLayer.cs:
```diff
@@ -0,0 +1,120 @@
+		protected override void WorldLoaded(World w, WorldRenderer wr)
+		{
+			// Cache locations of veinhole actors
+			w.ActorAdded -= ActorAddedToWorld;
+			w.ActorRemoved -= ActorRemovedFromWorld;
```
+=?

teinarss commented on OpenRA.Mods.Cnc/Traits/World/TSVeinsRenderer.cs:
```diff
@@ -0,0 +1,335 @@
+		{
+			// Cache locations of veinhole actors
+			// TODO: Add support for monitoring actors placed in the map editor!
+			w.ActorAdded -= ActorAddedToWorld;
+			w.ActorRemoved -= ActorRemovedFromWorld;
```
Same here?

pchote commented:
Updated.

Mailaender commented on OpenRA.Mods.Common/Traits/Buildings/Refinery.cs:
```diff
@@ -100,32 +103,42 @@ public IEnumerable<TraitPair<Harvester>> GetLinkedHarvesters()
 			}
 			else
-				amount = playerResources.ChangeCash(amount);
+				count = playerResources.ChangeCash(count);
```
should be `value`?

pchote replied:
🤦

pchote commented:
Fixed.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Refinery.cs:
```diff
@@ -44,6 +44,9 @@ public class RefineryInfo : TraitInfo, Requires<WithSpriteBodyInfo>, IAcceptReso
 		public readonly bool DiscardExcessResources = false;
 
+		[Desc("Value of each accepted resource type.", "Dictionary of [resource type]: [value per unit].")]
+		public readonly Dictionary<string, int> ResourceValues = new Dictionary<string, int>();
```
This is going to conflict with the logical next step, which is to convert resource storage from money to resources.

This should instead move to `PlayerResources`.

pchote commented:
Rebased and updated.

Moving `ResourceValues` from `Refinery` to `PlayerResources` had another nice side-effect of not requiring the values to be defined again on `EditorResourceLayer`, provided the mod defines them on the `^BasePlayer` (update rules have been updated to explain this).

teinarss commented on OpenRA.Mods.Cnc/Traits/World/TSVeinsRenderer.cs:
```diff
@@ -0,0 +1,335 @@
+			world = self.World;
+
+			resourceLayer = self.Trait<IResourceLayer>();
+			resourceLayer.CellChanged += AddDirtyCell;
```
Need to remove the eventhandler?

pchote commented:
Fixed.

teinarss commented:
Needs a rebase.

pchote commented:
Rebased.

pchote committed:
Implement proper TS Tiberium and Vein logic.

pchote commented:
Rebased and updated to account for the new sequence fields.

penev92 commented on OpenRA.Mods.Common/Traits/BotModules/BotModuleLogic/BaseBuilderQueueManager.cs:
```diff
@@ -418,15 +418,18 @@ ActorInfo ChooseBuildingToBuild(ProductionQueue queue)
-						if (found != null)
-							return found;
+						var nearbyResources = world.Map.FindTilesInAnnulus(baseCenter, baseBuilder.Info.MinBaseRadius, baseBuilder.Info.MaxBaseRadius)
+							.Where(a => resourceLayer.GetResource(a).Type != null)
```
If the ResourceLayer (and the AI and the Refinery) is to support multiple resources this would need to be a bit more specific than a null check.

pchote replied:
I agree, but this would require new AI plumbing and yaml options to specify which refinery types are associated with each resource type. The changes here are keeping the same behaviour the AI has always had with the new system. Improving this behaviour is a whole PR (maybe more than one) in its own right!

penev92 replied:
Yeah that was a general comment.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Refinery.cs:
```diff
@@ -86,7 +86,7 @@ public Refinery(Actor self, RefineryInfo info)
 		void INotifyCreated.Created(Actor self)
 		{
-			resourceMultipliers = self.TraitsImplementing<RefineryResourceMultiplier>().ToArray();
+			resourceValueModifiers = self.TraitsImplementing<IResourceValueModifier>().ToArray().Select(m => m.GetResourceValueModifier());
```
Redundant `.ToArray()` before `.Select()`.
Also caching the modifier values at this point negates the `IResourceValueModifier` implementation(s) being Conditional! So I think going back to the old version is the correct way.

pchote replied:
This is actually the standard pattern that we use for modifiers:

![Screenshot 2021-02-26 at 22 55 14](https://user-images.githubusercontent.com/167819/109363982-bbb33700-7885-11eb-8cfe-5063543a5c4b.png)

This allows us to keep the dynamic bit (evaluating the current modifiers, which does account for the trait being conditional!) dynamic, but saves the performance cost of redundantly querying the same set of traits each time.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Refinery.cs:
```diff
@@ -100,32 +100,42 @@ public IEnumerable<TraitPair<Harvester>> GetLinkedHarvesters()
-		public bool CanGiveResource(int amount) { return !info.UseStorage || info.DiscardExcessResources || playerResources.CanGiveResources(amount); }
-
-		public void GiveResource(int amount)
+		int IAcceptResources.GiveResource(string resourceType, int count)
```
Personal request: Since you're changing the method signature anyway, can we rename it to `AcceptResource` please? `Give` makes me think of exactly the opposite function it's performing, like it should be on `Harvester`, not `Refinery`.

penev92 commented on OpenRA.Mods.Common/Traits/Harvester.cs:
```diff
@@ -231,40 +228,42 @@ void UpdateCondition(Actor self)
+						continue;
+
+					contents[resourceType] -= accepted;
+					if (contents[resourceType] == 0)
```
This is tingling my spidey sense - I know it should be right, but we are still counting on the `Refinery` trait (or whatever `IAcceptResources` implementation is on the receiving side) to not give us a bogus number and with the subtraction above make us go in the negative (from where we can't recover). It would be prudent to either check for `<= 0`, or even throw an exception if we are in the negative and then keep this check as is.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Refinery.cs:
```diff
@@ -100,32 +100,42 @@ public IEnumerable<TraitPair<Harvester>> GetLinkedHarvesters()
 			}
 			else
-				amount = playerResources.ChangeCash(amount);
+				value = playerResources.ChangeCash(value);
```
Is `PlayerResources.ChangeCash()` (actually the entire `PlayerResources`?) supposed to change in the future?

pchote replied:
Yes.

I'm not sure where you were going with this, so i don't know how to give a more detailed answer 😅

penev92 replied:
I was going for "there really shouldn't be any `cash` involved". Come to think of it, cash should be a separate resource defined in the rules, not hardcoded in the engine 🤔  Then harvested resources can be converted to the "cash" resource and that can be used for purchasing stuff.

pchote replied:
That sounds similar to my end goal, but there are 2 or 3 PRs worth of work before we get there.

penev92 commented on OpenRA.Mods.Common/Traits/World/ResourceRenderer.cs:
```diff
@@ -18,60 +18,106 @@
 
 namespace OpenRA.Mods.Common.Traits
 {
+	public interface IResourceRenderer
```
This should probably go to `TraitsInterfaces.cs`.
Also probably could use a `RequireExplicitImplementationAttribute`.

penev92 commented on OpenRA.Mods.Common/Traits/Harvester.cs:
```diff
@@ -386,9 +385,9 @@ public bool CanTarget(Actor self, in Target target, List<Actor> othersAtTarget,
 
 				var info = self.Info.TraitInfo<HarvesterInfo>();
-				var res = self.World.WorldActor.TraitsImplementing<ResourceRenderer>()
+				var res = self.World.WorldActor.TraitsImplementing<IResourceRenderer>()
```
Is there a reason this is using `IResourceRenderer` and not `IResourceLayer`?

pchote replied:
`IResourceRenderer` accounts for the fog of war and player vision, `IResourceLayer` does not.

penev92 commented on OpenRA.Mods.Common/Traits/World/ResourceLayer.cs:
```diff
@@ -10,204 +10,308 @@
+			if (content.Type != resourceType)
+				return false;
+
+			return content.Density + amount <= resourceInfo.MaxDensity;
```
`AddResource` below would add regardles by capping at `MaxDensity`. This should probably be
```suggestion
			return content.Density <= resourceInfo.MaxDensity;
```

pchote replied:
I agree in theory, but in practice this would break resource spawning. I'm doing my best in this PR to work around badly designed legacy plumbing, without having to burn everything to the ground and restart from scratch.

penev92 commented on OpenRA.Mods.Common/Traits/SeedsResource.cs:
```diff
@@ -64,12 +55,11 @@ public void Seed(Actor self)
 				.Take(info.MaxRange)
-				.SkipWhile(p => !self.World.Map.Contains(p) ||
-					(resLayer.GetResourceType(p) == resourceType && resLayer.IsFull(p)))
+				.SkipWhile(p => resourceLayer.GetResource(p).Type == info.ResourceType && !resourceLayer.CanAddResource(info.ResourceType, p))
```
Are you sure we can drop the `Map.Contains(p)` check here? 🤔

pchote replied:
`resourceLayer.GetResource` and `resourceLayer.CanAddResource` now behave correctly when given a bogus cell, so the extra check here isn't needed.

penev92 commented on OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs:
```diff
@@ -13,40 +13,80 @@
+		ResourceLayerContents IResourceLayer.GetResource(CPos cell) { return Tiles.Contains(cell) ? Tiles[cell] : default; }
+		int IResourceLayer.GetMaxDensity(string resourceType)
+		{
+			if (!info.ResourceTypes.TryGetValue(resourceType, out var resourceInfo))
```
```suggestion
			return info.ResourceTypes.TryGetValue(resourceType, out var resourceInfo) ? resourceInfo.MaxDensity : 0;

```

penev92 commented on OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs:
```diff
@@ -13,40 +13,80 @@
+		[FieldLoader.LoadUsing(nameof(LoadResourceTypes))]
+		public readonly Dictionary<string, ResourceLayerInfo.ResourceInfo> ResourceTypes = new Dictionary<string, ResourceLayerInfo.ResourceInfo>();
+
+		protected static object LoadResourceTypes(MiniYaml yaml)
```
Please add a comment this was copied from `ResourceLayer` and `ResourceRenderer`.

penev92 commented on OpenRA.Mods.Common/Traits/World/ResourceRenderer.cs:
```diff
@@ -18,60 +18,106 @@
+		[FieldLoader.LoadUsing(nameof(LoadResourceTypes))]
+		public readonly Dictionary<string, ResourceInfo> ResourceTypes = new Dictionary<string, ResourceInfo>();
+
+		protected static object LoadResourceTypes(MiniYaml yaml)
```
Please add a comment this was copied from `ResourceLayer` and `EditorResourceLayer`.

penev92 commented on OpenRA.Mods.Common/Traits/World/ResourceLayer.cs:
```diff
@@ -10,204 +10,308 @@
+		public readonly bool RecalculateResourceDensity = false;
 
-		public ResourceLayer(Actor self)
+		protected static object LoadResourceTypes(MiniYaml yaml)
```
Please add a comment this was copied from `ResourceRenderer` and `EditorResourceLayer`.

penev92 commented on OpenRA.Mods.Common/Traits/World/ResourceLayer.cs:
```diff
@@ -10,204 +10,308 @@
+		}
+
+		[FieldLoader.LoadUsing(nameof(LoadResourceTypes))]
+		public readonly Dictionary<string, ResourceInfo> ResourceTypes = new Dictionary<string, ResourceInfo>();
```
This seems like an unnecessary initialization.

penev92 commented on OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs:
```diff
@@ -13,40 +13,80 @@
 	{
-		public override object Create(ActorInitializer init) { return new EditorResourceLayer(init.Self); }
+		[FieldLoader.LoadUsing(nameof(LoadResourceTypes))]
+		public readonly Dictionary<string, ResourceLayerInfo.ResourceInfo> ResourceTypes = new Dictionary<string, ResourceLayerInfo.ResourceInfo>();
```
This seems like an unnecessary initialization.

penev92 commented on OpenRA.Mods.Common/Traits/World/ResourceRenderer.cs:
```diff
@@ -18,60 +18,106 @@
+		}
+
+		[FieldLoader.LoadUsing(nameof(LoadResourceTypes))]
+		public readonly Dictionary<string, ResourceInfo> ResourceTypes = new Dictionary<string, ResourceInfo>();
```
This seems like an unnecessary initialization.

penev92 commented on OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs:
```diff
@@ -13,40 +13,80 @@
+		}
+
+		[Desc("Override the density saved in maps with values calculated based on the number of neighbouring resource cells.")]
+		public readonly bool RecalculateResourceDensity = false;
```
Can you elaborate on the purpose of this?
It seems fine for the `EditorResourceLayer` (if it's always on or there is a specific use-case)...

pchote replied:
This allows us to turn off the hacky legacy behaviour in TS without having to rework everything in the other mods (which is going to be a big and ugly PR in its own right).

penev92 commented on OpenRA.Mods.Common/Traits/World/ResourceLayer.cs:
```diff
@@ -10,204 +10,308 @@
 
-		public event Action<CPos, ResourceType> CellChanged;
+		[Desc("Override the density saved in maps with values calculated based on the number of neighbouring resource cells.")]
+		public readonly bool RecalculateResourceDensity = false;
```
Can you elaborate on the purpose of this?
It seems pretty out of place on the actual game world. Shouldn't this be a job for the Map Editor to calculate and store in the map so when going into a game the engine can just use it precalculated?

pchote replied:
TLDR: The way OpenRA has always dealt with resource densities is a mess, and fixing this properly is going to require a big and ugly PR in its own right (including changes to many maps and a new editor brush tool for adjusting resource densities).

This behaviour isn't compatible with the way TS's resources work, so this flag lets us fix the gameplay behaviour in TS. This regresses resource placement in the editor, which is an acceptable tradeoff for TS because the editor is not generally usable there yet.

penev92 commented on OpenRA.Mods.Common/Traits/World/ResourceLayer.cs:
```diff
@@ -10,204 +10,308 @@
+		{
+			[FieldLoader.Require]
+			[Desc("Resource index in the binary map data.")]
+			public readonly byte ResourceIndex = 1;
```
If this has `FieldLoader.RequireAttribute`, why the initialization (with a non-default value, even!) ?

penev92 commented on OpenRA.Mods.Common/Traits/World/ResourceLayer.cs:
```diff
@@ -10,204 +10,308 @@
 	{
-		public override object Create(ActorInitializer init) { return new ResourceLayer(init.Self); }
-	}
+		public class ResourceInfo
```
Maybe `ResourceTypeInfo` to better match the name of the property that uses it (and ofc the YAML representation)?

penev92 commented on OpenRA.Mods.Common/Traits/World/ResourceRenderer.cs:
```diff
@@ -18,60 +18,106 @@
-		[FieldLoader.Require]
-		[Desc("Only render these ResourceType names.")]
-		public readonly string[] RenderTypes = null;
+		public class ResourceInfo
```
Maybe `ResourceTypeInfo` to better match the name of the property that uses it (and ofc the YAML representation)?

penev92 commented on OpenRA.Mods.Common/Traits/World/ResourceRenderer.cs:
```diff
@@ -205,30 +232,35 @@ void INotifyActorDisposing.Disposing(Actor self)
+		protected virtual string GetRenderedResourceType(CPos cell) { return RenderContents[cell].Type; }
 
-		protected virtual string GetRenderedResourceTooltip(CPos cell) { return RenderContent[cell].Type?.Info.Name; }
+		protected virtual string GetRenderedResourceTooltip(CPos cell) { return RenderContents[cell].Info?.Name; }
```
Could this actually be null?

pchote replied:
Yes, if the cell doesn't contain any resources 😁

penev92 replied:
Oh I was expecting `RenderContents[cell]` to be `null` in that case.

penev92 commented on OpenRA.Mods.Common/UpdateRules/Rules/20201213/RemoveResourceType.cs:
```diff
@@ -0,0 +1,131 @@
+
+			if (resourceLayer.Nodes.Any())
+				yield return "Support for AllowUnderActors, AllowUnderBuildings, and AllowOnRamps have been removed.\n" +
+					"You must define a custom ResourceLayer subclass if you want to customize the default behaviour.";
```
Just to be clear: does this mean (considering the rest of the changes in this PR) we are effectively hardcoding it to spawn resources under actors, but not under buildings and not on ramps?

pchote replied:
Yes. These options were counterproductive because the actual rules that we need to implement are too arbitrary to encode in a generic flag (e.g. TS allows resources only on ramp types 0-4, but never 5-20, and veins requires several sentences to explain). I felt it would be better to hardcode a sensible neutral default and require all mods that want something different to subclass the layer than to implement flags for `IsTSTiberium` and `IsTSVeins` which would otherwise be needed.

penev92 commented on mods/cnc/rules/world.yaml:
```diff
@@ -126,30 +126,16 @@
+			BlueTiberium:
+				Sequences: bti1, bti2, bti3, bti4, bti5, bti6, bti7, bti8, bti9, bti10, bti11, bti12
+				Palette: bluetiberium
+				Name: Tiberium
```
Since this is for tooltips, do we not want to distinguish between "Tiberium" and "Blue Tiberium"?
Also might be better to rename the property to `Tooltip` and potentially add a separate `Name` property if needed.

penev92 commented on mods/ts/rules/world.yaml:
```diff
@@ -198,47 +198,39 @@
+			BlueTiberium:
+				Sequences: tib01, tib02, tib03, tib04, tib05, tib06, tib07, tib08, tib09, tib10, tib11, tib12
+				Palette: bluetiberium
+				Name: Tiberium
```
Since this is for tooltips, do we not want to distinguish between "Tiberium" and "Blue Tiberium"?
Also might be better to rename the property to `Tooltip` and potentially add a separate `Name` property if needed.

pchote replied:
We don't, no. Why would we?

penev92 replied:
I think at least lore-wise they are different 🤷

penev92 commented on OpenRA.Mods.Cnc/Traits/World/TSEditorResourceLayer.cs:
```diff
@@ -0,0 +1,110 @@
+			if (resourceType == info.VeinType)
+			{
+				var neighboursValid = Common.Util.ExpandFootprint(cell, false)
+					.All(c => IsValidVeinNeighbour(cell, c));
```
Is this checking if all non-diagonal neightbours could potentially (in time) contain veins?

pchote replied:
This is the bit that handles

> The original TS engine implemented resources as overlays (along with walls, bridges, crates, etc), which meant that harvestable resources were automatically not allowed to overlap with visual only things. Veins were a bit weird in that the LAT border cells were not themselves harvestable (they gave a harvest cursor but the harvester would go to the closest interior cell), which means that our `ResourceLayer` code needs to reserve an empty cell between veins and tiberium so that the `ResourceRenderer` can draw the borders without it overlapping with tiberium. 

when painting veins in the editor.

penev92 replied:
So because we need to have a visual-only border around the actual vein cells, all neighbours of vein cells must also be valid vein cells for the illusion to work? 🤔

pchote replied:
Non-diagonal neighbours, yeah.

penev92 commented on OpenRA.Mods.Cnc/Traits/World/TSEditorResourceLayer.cs:
```diff
@@ -0,0 +1,110 @@
+			}
+
+			// TODO: Check against actors in the EditorActorLayer
+			return true;
```
Shouldn't we check for buildings at least?

pchote replied:
Yes, but this PR is already too big without trying to fix all of the long-standing missing editor features at once 😅 

(This comment is documenting existing misbehaviour)

penev92 commented on OpenRA.Mods.Cnc/Traits/World/TSTiberiumRenderer.cs:
```diff
@@ -0,0 +1,92 @@
+				default: variants = Variants; break;
+			}
+
+			return variants[resourceType].Values.Random(world.LocalRandom);
```
Base class uses `Game.CosmeticRandom` here.

penev92 commented on OpenRA.Mods.Cnc/Traits/World/TSVeinsRenderer.cs:
```diff
@@ -0,0 +1,338 @@
+		void UpdateSpriteLayers(CPos cell, int[] indices)
+		{
+			if (indices != null)
+				spriteLayer.Update(cell, veinSequence, veinPalette, indices.Random(world.LocalRandom));
```
~~Base class~~ `ResourceRenderer` uses `Game.CosmeticRandom` here.

penev92 commented on mods/ts/rules/world.yaml:
```diff
@@ -198,47 +198,39 @@
+				Sequences: tib01, tib02, tib03, tib04, tib05, tib06, tib07, tib08, tib09, tib10, tib11, tib12
+				Palette: bluetiberium
+				Name: Tiberium
+		Ramp1Sequences:
```
For this setup to be consistent with itself, these should really be defined on each ResourceType. Probably overkill for such a non-generic trait though.

pchote replied:
`ResourceTypes` is inherited from `ResourceLayer` so can't be overridden without introducing massive code duplication. Unless i'm missing something obvious?

penev92 replied:
`TSTiberiumRenderer(Info)` inherits the property `ResourceTypes` (of type `ResourceRendererInfo.ResourceInfo`) from `ResourceRenderer(Info)`. Not to be confused with the `ResourceTypes` property of `ResourceLayerInfo.ResourceInfo` on `ResourceLayerInfo`. 😄 
Apart from that, it should be doable to override said property and also add a `public class TiberiumInfo : ResourceInfo` inside `TSTiberiumRendererInfo`, but like I said - very likely not worth in this particular case.

pchote replied:
It would be nice if it were this simple, but `ResourceTypes` is a field that it magically populated using a static method, so there's no room for virtual/overrides. We could use `new` to hide the base implementation, but then we would be parsing the resource types twice and different traits would be using different versions which I think is asking for trouble.

pchote replied:
The one thing we could do is to split out an abstract `ResourceRendererBase` trait / traitinfo that both inherit individually. This adds a bit more duplication and complexity, but perhaps justified here?

penev92 replied:
It would definitely be an improvement to do so. I don't know how useful the added flexibility would be in practice though (for community mods ofc, not for our default ones). So up to you.

pchote replied:
I can't find a way to implement this that wouldn't be a total nightmare, so will skip it.

pchote commented:
~There are a bunch of properties that should now also change to expression bodied syntax.~

pchote committed:
Add IResourceRenderer interface.

pchote committed:
Replace ResourceLayer references with IResourceLayer in traits/warheads.

pchote committed:
Overhaul resource layer logic:

* ResourceType trait has been removed.
* Simulation-related data is now defined on the
  ResourceLayer (which mods can subclass/replace).
* Support non-money resources by moving the resource
  values to the PlayerResources trait.
* Allow mods to disable the neighbour density override
  and instead always use the map-defined densities.
* Allow mods to define their own resource placement
  logic (e.g. allow resources on slopes) by subclassing
  (Editor)ResourceLayer.
* Improve ability to subclass/override ResourceRenderer
  by exposing more virtual methods.

pchote committed:
Pass resource type and count to IAcceptResources instead of the value.

pchote commented:
Fixed everything i reasonably can.

abcdefg30 commented:
Unfortunately there are merge conflicts now.

pchote commented:
Rebased.

reaperrr reviewed (approved):
LGTM

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d3bd82e2ff58e811cea5eae47821a39cca1b0c8d...d4f6a1caf5d32bbf013f42492daa2b5150e3cb2b
